syncserver/src/server/mod.rs

```diff
@@ -371,3 +396,92 @@ pub fn build_cors(settings: &Settings) -> Cors {
 
     cors
 }
+
+/// Emit database pool and threadpool metrics periodically
+fn spawn_metric_periodic_reporter<T: GetPoolState + Send + 'static>(
+    interval: Duration,
+    metrics: StatsdClient,
+    pool: T,
+    blocking_threadpool: Arc<BlockingThreadpool>,
+) -> Result<(), DbError> {
+    let hostname = hostname::get()
+        .expect("Couldn't get hostname")
+        .into_string()
+        .expect("Couldn't get hostname");
+    let blocking_threadpool_size =
+        str::parse::<u64>(&env::var("ACTIX_THREADPOOL").unwrap()).unwrap();
+    tokio::spawn(async move {
+        loop {
+            let PoolState {
+                connections,
+                idle_connections,
+            } = pool.state();
+            metrics
+                .gauge_with_tags(
+                    "storage.pool.connections.active",
+                    (connections - idle_connections) as u64,
+                )
+                .with_tag("hostname", &hostname)
+                .send();
+            metrics
+                .gauge_with_tags("storage.pool.connections.idle", idle_connections as u64)
+                .with_tag("hostname", &hostname)
+                .send();
+
+            let active_threads = blocking_threadpool.active_threads();
+            let idle_threads = blocking_threadpool_size - active_threads;
+            metrics
+                .gauge_with_tags("blocking_threadpool.active", active_threads)
+                .with_tag("hostname", &hostname)
+                .send();
+            metrics
+                .gauge_with_tags("blocking_threadpool.idle", idle_threads)
+                .with_tag("hostname", &hostname)
+                .send();
+
+            time::delay_for(interval).await;
+        }
+    });
+
+    Ok(())
+}
+
+/// A threadpool on which callers can spawn non-CPU-bound tasks that block their thread (this is
+/// mostly useful for running I/O tasks). `BlockingThreadpool` intentionally does not implement
+/// `Clone`: `Arc`s are not used internally, so a `BlockingThreadpool` should be instantiated once
+/// and shared by passing around `Arc<BlockingThreadpool>`s.
+#[derive(Debug, Default)]
+pub struct BlockingThreadpool {
+    spawned_tasks: AtomicU64,
+}
+
+impl BlockingThreadpool {
+    /// Runs a function as a task on the blocking threadpool.
+    ///
+    /// WARNING: Spawning a blocking task through means other than calling this method will
+    /// result in inaccurate threadpool metrics being reported. If you want to spawn a task on
+    /// the blocking threadpool, you **must** use this function.
+    pub async fn spawn<F, T, E>(&self, f: F) -> Result<T, E>
+    where
+        F: FnOnce() -> Result<T, E> + Send + 'static,
+        T: Send + 'static,
+        E: fmt::Debug + Send + InternalError + 'static,
```

Comment:
Requiring `E: InternalError` prevents callers from having to pass a function manually to construct `E`'s version of an internal error